// @monodon/typescript-nx-imports-plugin/src/lib/nx-imports-plugin.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.NxImportsPlugin = void 0;
const path_1 = require("path");
const isNxImportPlugin = Symbol('__isNxImportPlugin__');
class NxImportsPlugin {
    constructor(typescript) {
        this.typescript = typescript;
        this.config = {};
        this.projects = new Map();
    }
    setConfig(config) {
        this.logger?.log('setting configuration ' + JSON.stringify(config));
        this.config = config;
        this.projects.forEach((project) => {
            this.updateProject(project);
        });
    }
    addProject(project) {
        this.logger?.log('addProject ' + project.getProjectName());
        if (this.projects.has(project.getProjectName())) {
            this.logger?.log('project already tracked ' + project.getProjectName());
            return;
        }
        this.projects.set(project.getProjectName(), project);
        this.updateProject(project);
    }
    updateProject(project) {
        this.logger?.log('updating project: ' + project.getProjectName());
        const externals = this.getRootFiles(project);
        externals.forEach((external) => {
            project.addMissingFileRoot(this.typescript.server.toNormalizedPath(external));
        });
    }
    getRootFiles(project) {
        this.logger?.log('get root files: ' + JSON.stringify(this.config));
        const externalFiles = this.config.externalFiles || [];
        if (externalFiles.length === 0) {
            return [];
        }
        const projectDirectory = (0, path_1.dirname)(project.getProjectName());
        this.logger?.log(`project directory: ${projectDirectory}`);
        const filteredExternalFiles = externalFiles
            .filter(({ directory }) => {
            return !projectDirectory.startsWith(directory);
        })
            .map(({ mainFile }) => mainFile);
        this.logger?.log(`root files: ${JSON.stringify(filteredExternalFiles)}`);
        return filteredExternalFiles;
    }
    decorate(languageService) {
        this.logger?.log('decorate');
        if (languageService[isNxImportPlugin]) {
            // Already decorated
            return;
        }
        const intercept = Object.create(null);
        const oldGetCompletionsAtPosition = languageService.getCompletionsAtPosition.bind(languageService);
        intercept.getCompletionsAtPosition = (fileName, position, options) => {
            this.logger?.log(`getCompletionsAtPosition ${fileName}:${position}`);
            return this.getCompletionsAtPosition(oldGetCompletionsAtPosition, fileName, position, options);
        };
        return new Proxy(languageService, {
            get: (target, property) => {
                if (property === isNxImportPlugin) {
                    return true;
                }
                return intercept[property] || target[property];
            },
        });
    }
    getCompletionsAtPosition(delegate, fileName, position, options) {
        return delegate(fileName, position, options);
    }
}
exports.NxImportsPlugin = NxImportsPlugin;
